### PR TITLE
Combination of a few Network class enhancements, particularly to support partitioning better.

### PR DESCRIPTION
**Pull request by mattw-nws**
[Short description explaining the high-level reason for the pull request]

## Additions

-

## Removals

-

## Changes

-

## Testing

1.

## Notes

-

## Todos

-

## Checklist

- [x] PR has an informative and human-readable title
- [x] Changes are limited to a single goal (no scope creep)
- [x] Code can be automatically merged (no conflicts)
- [x] Code follows project standards (link if applicable)
- [x] Passes all existing automated tests
- [x] Any _change_ in functionality is tested
- [x] New functions are documented (with a description, list of inputs, and expected output)
- [x] Placeholder code is flagged / future todos are captured in comments
- [x] Project documentation has been updated (including the "Unreleased" section of the CHANGELOG)
- [x] Reviewers requested with the [Reviewers tool](https://help.github.com/articles/requesting-a-pull-request-review/) :arrow_right:

## Testing checklist (automated report can be put here)

1.

### Target Environment support

- [ ] Linux

### CODE REVIEW

File: include/core/network.hpp [hunk 1]
```diff
@@ -35,15 +77,28 @@ namespace network {
...
           //which is a reasonable semantic
           return topo_order | boost::adaptors::reversed
                             | boost::adaptors::transformed([this](int const& i) { return get_id(i); })
                             | boost::adaptors::filtered([type](std::string const& s) { return s.substr(0,3) == type; });
         }
+        auto filter_dfr(std::string type){
+          // This would be more semantically expected to me:
+          //   return get_dfr_from_headwaters() | boost::adaptors::reversed ...
+          // ...but it messes up the NetworkIndexT vector somehow.
+          // Try it and run the test, we get indexes that don't exist when iterating.
```
[the remark this answers is not in the document]
Comment (hellkite500) [reply include/core/network.hpp:92]: Also, probably best to use a const reference
```c++
NetworkIndexT const & get_dfr_from_headwaters();
```

**Comment by mattw-nws** [comment]
Implmenting subsetting or subgraphs in or near the Network class is probably valuable, but a bit involved. For the immediate timeframe, recommend doing this in [partitionGener](https://github.com/NOAA-OWP/ngen/blob/a1e4bcd3f589aa11944d9c139e97556f1c9c5224/src/partitionGenerator.cpp#L433):

PartitionGenerator--in the near term, rather than geojson::read(...) inside the loop, do the following:

1. Make a duplicate of `global_nexus_collection` using the FeatureCollection copy constructor (geojson::read returns a 
   FeatureCollection, which has what looks like a cheap copy constructor).
  a. The `global_nexus_collection` at the point the loop starts has both nexi AND catchments. Recommend calling the copy 
     `local_feature_collection`, perhaps.
2. Loop through `global_nexus_collection` (yes, the original), and:
  a. For any catchment that is not in the partition (that is, in the `local_cat_ids` vector), call 
     `remove_feature_by_id` on the COPY.
  b. Remove any nexus that is not attached to a catchment in the partition (less clear how to accomplish this for 
     upstream nexi...?)
3. Make a network `local_network` from the COPY.

Rinse and repeat for each partition.

Note that the current code appears to try to create a network from catchments only, which will fail. 
Unless `geojson::read(file_path, ids)` is automatically including related nexi attached to catchment IDs,
which doesn't appear to be so.

**Comment by hellkite500** [comment]
> Note that the current code appears to try to create a network from catchments only, which will fail.
> Unless `geojson::read(file_path, ids)` is automatically including related nexi attached to catchment IDs,
> which doesn't appear to be so.

`Network local_network(local_catchment_collection, &link_key);` This will use the `link_key` to look up the nexus from feature properties of the catchment feature.

**Comment by hellkite500** [comment]
> 2. Loop through `global_nexus_collection` (yes, the original), and:
>    a. For any catchment that is not in the partition (that is, in the `local_cat_ids` vector), call
>    `remove_feature_by_id` on the COPY.
>    b. Remove any nexus that is not attached to a catchment in the partition (less clear how to accomplish this for
>    upstream nexi...?)

I think it would be simpler to add a function to the feature collection to filter/subset the collection.  Essentially the same concept, but put into the feature collection, so the caller does this on the copied collection:
```c++
local_feature_collection.filter(local_cat_ids)
```
As far as the semantics of removing a feature, making an overload of the filter function with a flag to remove linked features may be the best route.
```c++
local_feature_collection.filter(local_cat_ids, True)
```
would remove all features with `cat_id` in `local_cat_ids` as well remove immediately linked (destination) features.

**Comment by mattw-nws** [comment]
> `Network local_network(local_catchment_collection, &link_key);` This will use the `link_key` to look up the nexus from feature properties of the catchment feature.

Yes, but I'm referring to the [earlier creation of the global `catchment_network` object](https://github.com/NOAA-OWP/ngen/blob/master/src/partitionGenerator.cpp#L381], which is used to pass to `generate_partitions` directly below. That's done before the [`nexusDataFile` is even read](https://github.com/NOAA-OWP/ngen/blob/master/src/partitionGenerator.cpp#L390), so I don't believe the nexus features are in the collection, so `generate_partitions` will operate on a completely disconnected network.

**Comment by hellkite500** [comment]
> Yes, but I'm referring to the [earlier creation of the global `catchment_network` object](https://github.com/NOAA-OWP/ngen/blob/master/src/partitionGenerator.cpp#L381), which is used to pass to `generate_partitions` directly below. That's done before the [`nexusDataFile` is even read](https://github.com/NOAA-OWP/ngen/blob/master/src/partitionGenerator.cpp#L390), so I don't believe the nexus features are in the collection, so `generate_partitions` will operate on a completely disconnected network.

The same applies to this network, it reads the nexus identities from the catchment features using the `link_key`

**Comment by mattw-nws** [comment]
@donaldwj I believe this branch now traverses the tree as you expected...if you can checkout this branch and try it I'd appreciate that.

Next will try to speed up/eliminate that ::read inside the loop.

**Comment by hellkite500** [comment]
It seems the merge commit makes the change set here include all the files that were included in that merge (that have already been reviewed and merged.) I'm not sure what your exact git flow is, but you might want to review the [git flow document](https://github.com/NOAA-OWP/ngen/blob/master/doc/GIT_USAGE.md#getting-upstream-changes) for ngen, specifically the part about syncing upstream and keeping forks up to date.

Also, if the merges are a result of doing a `git pull`, you might simply consider using `git pull --rebase` to maintain the linear history.

**Comment by mattw-nws** [comment]
@hellkite500 Yes... this didn't look right. I'm relatively new to rebasing workflow, but I did rebase from upstream/master... Not sure what happened there. Probably wrong thing since this was branched from my fork/master....

**Comment by mattw-nws** [comment]
> I believe this branch now traverses the tree as you expected

However...I'm fairly sure the determination of remote nexi is wrong. Though the selected catchments should be right now.

**Comment by donaldwj** [comment]
What makes you think finding the remotes is now wrong. I don't think that loop can break so long as the local and global networks are correctly formed, as it simply finds the nexi that are referenced in a local network but a member of it.

**Comment by mattw-nws** [comment]
> so long as the local and global networks are correctly formed

That's exactly it--if you trace `nexus_part` it is not added to the filter passed to `geojson::read` so no nexi are part of the constructed local network, as near as I can tell.

**Comment by hellkite500** [comment]
> That's exactly it--if you trace `nexus_part` it is not added to the filter passed to `geojson::read` so no nexi are part of the constructed local network, as near as I can tell.
@mattw-nws I think there is some confusion on how/where nexus can be found.  In the `catchment_data.geojson` each catchment features has a reference to its downstream nexus using property `toid`
```geojson
{ "type": "Feature", "id": "cat-3872", "properties": { "areasqkm": 18.157283748787744, "toid": "nex-0" }, "geometry": { "type": "MultiPolygon", "coordinates": [ ... ] } }
```
The local catchment collection
```c++
        // make a local network
        Network local_network(local_catchment_collection, &link_key);
```
is created without fully linking the nexus from the nexus feature collection, but it still has nexus references to each catchment's downstream nexus because it passes the link key to the constructor, which invokes the following to create the network object.

```c++
Network::Network( geojson::GeoJSON features, std::string* link_key = nullptr ){

  std::string feature_id, downstream_id;
  Graph::vertex_descriptor v1, v2;

  //TODO ensure all features are the same logical HY_Features type?
  for(auto& feature: *features)
  {
    feature_id = feature->get_id();
    if( this->descriptor_map.find( feature_id ) == this->descriptor_map.end() )
    {
      //Haven't visited this feature yet, add it to graph
      //add vertex to graph
      v1 = add_vertex( feature_id, this->graph );
      this->descriptor_map.emplace( feature_id, v1 );
    }
    else{
      v1 = this->descriptor_map[ feature_id ];
    }

      if (link_key != nullptr and feature->has_property(*link_key)) {

          downstream_id = feature->get_property(*link_key).as_string();
          if( this->descriptor_map.find(downstream_id) != this->descriptor_map.end() )
          {
            v2 = this->descriptor_map[ downstream_id ];
          }
          else {
            v2 = add_vertex( downstream_id, this->graph );
            this->descriptor_map.emplace( downstream_id, v2 );
          }
            add_edge(v1, v2, this->graph);
      }
  }

  init_indicies();

}
```

Note particularly these lines:
```c++
 if (link_key != nullptr and feature->has_property(*link_key)) {

          downstream_id = feature->get_property(*link_key).as_string();
          if( this->descriptor_map.find(downstream_id) != this->descriptor_map.end() )
          {
            v2 = this->descriptor_map[ downstream_id ];
          }
          else {
            v2 = add_vertex( downstream_id, this->graph );
            this->descriptor_map.emplace( downstream_id, v2 );
          }
            add_edge(v1, v2, this->graph);
      }
```

So as long as the local collection doesn't need to find any upstream references from CATCHMENTS, then this works as expected.  And looking at the usage of `local_network`, the NEXUS objects reference `origination_ids`, but the catchment iterator is never used to query upstream/origination features.

**Comment by mattw-nws** [comment]
Still not convinced. Though I'm probably wrong, lets see.

This line:
```
generate_partitions(global_network, num_partitions, num_catchments, catchment_part, nexus_part);
```
populates nexi into `nexus_part`... those nexi are _part_ of the partition... they are the "local" nexi... yes? Then, this line gets a _different_ set of nexi and calls them local:
```
auto local_nexuses = local_network.filter("nex");
```
Here's one scenario where those will not match: If the partitioning process is crawling upstream (which is what my change accomplishes) and stops at a nexus, that nexus will be upstream of all local catchments and downstream of none. If the `local_network` contains only the catchments, that upstream nexus that was part of the partition will not be known at all in `local_network`. But it was included as part of the partition. Which also means it was _not included as _local_ in any other partition.

I have yet to trace through the rest in detail or try to run tests to verify, but I have a feeling that that mismatch is going to lead to nexi which are local in no partitions.

Here's one scenario where I think this fails...My understanding is that the partition may include an up

**Comment by mattw-nws** [comment]
Okay...phew... that seems to have cleaned up my history.

**Review by donaldwj — APPROVED** [review]
Testing with this results in partition files without the errors partition errors generated in the pervious versions. I would recommend a manual check of partition files for Sugar Creek before merging.

**Review by donaldwj — APPROVED** [review]

**Comment by mattw-nws** [comment]
@donaldwj  per your suggestion I ran this partitionGenerator on HUC01 with 16 and 256 partitions... it seems to have worked, at least it didn't crash. Suggestion, it may be relatively easy to add remote-symmetry checking to the [partition_labeller.py](https://github.com/NOAA-OWP/ngen/blob/0ceb89c67c24f2baf3b705ffc067c65f2331c489/utilities/partitioning/partition_labeler.py) script in this PR.

**Review by hellkite500 — APPROVED** [review]

**Pull request merged**